README
NAME
    Queue::Q - Mix-and-match Queue Implementations and Backends

SYNOPSIS
      # Pick you queue type and a back-end:
  
      # Very primitive FIFO queue (abstract interface)
      use Queue::Q::NaiveFIFO;
      # ... and its pure-Perl, in-memory implementation
      use Queue::Q::NaiveFIFO::Perl;
      # ... its Redis-based implementation
      use Queue::Q::NaiveFIFO::Redis;
  
      # "Reliable" enqueue, claim, mark-as-done FIFO queue:
      use Queue::Q::ClaimFIFO; # abstract interface
      # ... and again, its pure-Perl, in-memory implementation
      use Queue::Q::ClaimFIFO::Perl;
      # ... and its Redis-based implementation
      use Queue::Q::ClaimFIFO::Redis;
  
      # A composite, distributed queue, built from shards of
      # the above types with weak global and strict local
      # ordering, scaling beyond single nodes.
      use Queue::Q::DistFIFO; 
  
      # Usage is somewhat similar for most queue types:
      my $queue = Queue::Q::ClaimFIFO::Redis->new(
          server      => 'redis-01',
          port        => '6379',
          queue_name  => "image_queue"
      );
  
      $q->enqueue_item( $img_url );
  
      # In another process on another machine:
      while (defined( my $item = $q->claim_item )) {
          # Do some work!
          $q->mark_item_as_done($item);
      }

      # n-ary versions exist, too:
      $q->enqueue_items( @img_urls );
      my @items = $q->claim_items(10); # 10 at a time, padded with undef
      $q->mark_items_as_done(@items);

DESCRIPTION
    This is an experimental module. The interface may change without notice.
    Before using it in production, please get in touch with the authors!

    "Queue::Q" is a collection of queue implementations each with multiple
    backends. Right now, it comes with three basic queues:

    Queue::Q::NaiveFIFO
      A naive FIFO queue "NaiveFIFO". Supports enqueuing data and later
      claiming it in the order it was enqueued. Strict ordering,
      low-latency, high-throughput, no resilience against crashing workers.

    Queue::Q::ClaimFIFO
      A claim/mark-as-done FIFO queue "ClaimFIFO". Supports enqueuing data
      (as Queue::Q::ClaimFIFO::Item objects), claiming items, and keeping
      track of items-being-worked-on until they are reported as done. Strict
      ordering, slightly higher latency and lower throughput than the naive
      FIFO queue. Resilience against crashing workers when combined with a
      clean-up script that checks for old, claimed items.

    Queue::Q::ReliableFIFO
      Similar interface to the "ClaimFIFO" queue type, but only comes with a
      Redis-backed implementation now. Much faster for the basic operations
      then "ClaimFIFO" since it doesn't always have to execute Lua scripts.
      See Queue::Q::ReliableFIFO for details.

    The first two basic queues come with two back-end implementations each
    right now: One is a very simple, in-memory, single-process
    implementation Queue::Q::NaiveFIFO::Perl and Queue::Q::ClaimFIFO::Perl
    respectively. The other is an implementation based on Redis:
    Queue::Q::NaiveFIFO::Redis and Queue::Q::ClaimFIFO::Redis.

    As noted above, the "ReliableFIFO" queue has a Redis-based backend only
    for now.

    In addition to the basic queue types, the distribution contains
    Queue::Q::DistFIFO, an implementation of a distributed queue that can
    use the basic queues as shards (but only one type of shard per
    distributed queue). It supports all operations of the basic queues, but
    does not enforce strict global ordering, but weak global and strict
    local ordering. In an early test, two mid-range servers running multiple
    instances of Redis each sustained 800k-1M transactions per second with a
    naive-type distributed queue. "DistFIFO" has not been tested with the
    "ReliableFIFO" implementation as building blocks yet!

ACKNOWLEDGMENT
    This module was originally developed for Booking.com. With approval from
    Booking.com, this module was generalized and put on CPAN, for which the
    authors would like to express their gratitude.

AUTHOR
    Herald van der Breggen, <herald.vanderbreggen@booking.com>

    Steffen Mueller, <smueller@cpan.org>

COPYRIGHT AND LICENSE
    Copyright (C) 2012, 2013 by Steffen Mueller

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself, either Perl version 5.8.1 or, at
    your option, any later version of Perl 5 you may have available.



Changes
Revision history for Perl extension Queue::Q.

0.29
  - Migrate fifo-cli to Term::ReadLine

0.28
  - Bring Queue::Q::ClaimFIFO::Redis in line with latest Redis::ScriptCache
    changes

0.27
  - Add option warn_on_requeue to Queue::Q::ReliableFIFO::Redis to
    make warnings introduced in 0.24 optional by default

0.26
  - Fix uninitialized warnings in the redis implementation of ReliableFIFO

0.25
  - Add cp command to the CLI (Andrey Frolov)
  - Add new parameter ignore_config_file to run the CLI

0.24
  - don't modify the $options hash passed to ->consume
  - ReliableFIFO: Serialize JSON to binary octets instead of unicode codepoints.
  - Throw a warning when items are not properly removed from the 'busy'
    queue when using the Chunk => 1 option (which is the default one).
  - Lua: Make some error messages a bit more accurate

0.23
  - Repair broken CPAN upload

0.22
  - DieOnError option deprecated and renamed to ReturnOnDie
  - Better lua error messages
  - Disallow undefined server names or ports

0.21 Wed Oct 16 08:38:12 CEST 2013
  - bugfix introduced in 0.20 with requeue_failed_items when called without parameters

0.20 Fri Oct 11 14:13:06 CEST 2013
  - bugfixes for solving issues with very large number of failed items

0.19 Wed Aug 14 16:44:45 CEST 2013
  - bugfix to have ReliableFIFO::Redis working again in non blocking mode

0.18 Thu Aug  8 21:48:00 CEST 2013
  - Non-blocking ReliableFIFO::Redis claim_item_nonblocking
  - ReliableFIFO::Redis: New option NoSigHandlers for the consume() method
 
0.17 Fri May 24 13:27:47 CEST 2013
  - bugfix with MaxItems in ReliableFIFO/Redis
  - merged branch of ksurrent fixing inconsistent API for ClaimFIFO

0.16 Wed May  8 17:23:59 CEST 2013
  - added peek_item

0.15 Tue Apr  9 10:28:19 CEST 2013
  - added clone() method
  - fixed cleanup bug with handle_expired_items
  - fixed typo's in documentation

0.14 Wed Mar 20 14:34:03 CET 2013
  - solved bug with stop signal while handling a chunk of items

0.13 Tue Mar 19 10:26:26 CET 2013
  - added MaxSeconds option for consume()
  - added t_created for Item objects, while default timestamp reflects the 
    time the item was put in the queue (updated when requeuing).

0.12 Mon Feb 18 12:01:34 CET 2013
  - added option ReturnWhenEmpty, back to same defaults
  - rolled back fix for unclaiming

0.11 Fri Feb 15 16:00:00 CET 2013
  - Added two options for get_and_flush_failed_items()
  - Fix fifo-cli for non-system Perl usage
  - Fix claim_wait_timeout being ignored in consume
  - Attempt to fix bug with unclaiming/requeueing
    __requeue_busy: used by unclaim, requeue_busy_item, and requeue_busy_error

0.10 Fri Feb  1 16:40:49 CET 2013
  - solved bug with negative sleep time

0.09 Thu Jan 31 17:12:11 CET 2013
  - removed bug with handle_expired_items()
  - added requeue_busy_item() and requeue_failed_item()

0.08 Wed Jan 23 11:41:49 CET 2013
  - limit chunk size in claim_item(s) to queue length 
    for ReliableFIFO and NaiveFIFO
  - mark_as done more efficient in RelaibleFIFO (lrem starts
    at the right side of the list).

0.07 Tue Jan 22 15:12:04 CET 2013
  - fifo-cli change

0.06 Tue Jan 22 11:37:25 CET 2013
  - added get_and_flush_failed_items()
  - added added option Pause and ProcessAll for consume()

0.05 Thu Dec 13 20:38:45 CET 2012
  - MaxItems option for consume()

0.04 Tue Dec 11 11:45:09 CET 2012
  - EXEC_FILES correction in Makefile.PL

0.03 Mon Dec 10 20:03:48 CET 2012
  - added raw_items_* and memory_suage_perc() methods
    to Queue:Q::ReliableFIFO::Redis

0.02  Tue Dec  4 10:00:00 2012
  - added age() method to Queue:Q::ReliableFIFO::Redis

0.01  Mon Dec  3 16:00:00 2012
  - original version



Makefile.PL
use 5.008001;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'Queue::Q',
    VERSION_FROM      => 'lib/Queue/Q.pm', # finds $VERSION
    EXE_FILES         => [qw(scripts/fifo-cli)],
    PREREQ_PM         => {
      'Redis' => '1.955',
      'Redis::ScriptCache' => '0.01',
      'Sereal::Encoder' => '0.10',
      'Sereal::Decoder' => '0.10',
      'Class::XSAccessor' => '0.12',
      'List::Util' => '0',
      'Scalar::Util' => '0',
      'Time::HiRes' => '0',
      'constant' => '0',
      'Carp' => '0',
      'Digest::SHA1' => '0',
      'parent' => '0',
      'File::Spec' => '0',
      'Test::More' => '0.88',
      'JSON::XS' => '0',
      'File::Slurp' => '0',
    }, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/Queue/Q.pm', # retrieve abstract from module
       AUTHOR         => 'Steffen Mueller <smueller@cpan.org>') : ()),
);
